src/raftstore/store/snap/reader.rs

```diff
@@ -0,0 +1,274 @@
+                    e => return e,
+                }
+            }
+            self.cf_index += 1;
+            self.cf_file = None;
+        }
+        Ok(0)
+    }
+}
+
+pub(super) struct SnapshotLoader(Inner);
+
+impl SnapshotLoader {
+    pub(super) fn new(
+        dir: String,
+        for_send: bool,
+        key: SnapKey,
+        snapshot_meta: SnapshotMeta,
+    ) -> Self {
+        let inner = Inner::new(dir, for_send, key, snapshot_meta);
+        SnapshotLoader(inner)
+    }
+
+    pub(super) fn load(mut self) -> Result<SnapshotMeta> {
+        for cf in self.0.snapshot_meta.get_cf_files() {
+            if cf.get_size() == 0 {
+                continue;
+            }
+            let (dir, for_send, key) = (&self.0.dir, self.0.for_send, self.0.key);
+            let cf_file_path = get_cf_file_path(dir, for_send, key, cf.get_cf());
+
+            let checksum = calc_crc32(&cf_file_path)?;
+            if checksum != cf.get_checksum() {
+                return Err(snapshot_checksum_corrupt(cf.get_checksum(), checksum));
+            }
+
+            let size = get_file_size(&cf_file_path)?;
+            if size != cf.get_size() {
+                return Err(snapshot_size_corrupt(cf.get_size(), size));
+            }
+        }
+        Ok(mem::replace(&mut self.0.snapshot_meta, SnapshotMeta::new()))
+    }
+}
+
+pub(super) struct SnapshotApplyer {
+    inner: Inner,
+    snap_stale_notifier: Arc<SnapStaleNotifier>,
+    applied: Arc<AtomicBool>,
+}
+
+impl SnapshotApplyer {
+    pub(super) fn new(
+        dir: String,
+        key: SnapKey,
+        snapshot_meta: SnapshotMeta,
+        snap_stale_notifier: Arc<SnapStaleNotifier>,
+        applied: Arc<AtomicBool>,
+    ) -> Self {
+        let inner = Inner::new(dir, false, key, snapshot_meta);
+        SnapshotApplyer {
+            inner,
+            snap_stale_notifier,
+            applied,
+        }
+    }
+
+    pub(super) fn apply(&self, options: ApplyOptions) -> Result<()> {
+        for cf in self.inner.snapshot_meta.get_cf_files() {
+            if cf.get_size() == 0 {
+                continue;
+            }
+            let cf = cf.get_cf();
+            if plain_file_used(cf) {
+                self.apply_plain_cf_file(cf, &options)?;
+            } else {
+                let (dir, key) = (&self.inner.dir, self.inner.key);
+                let cf_path = get_cf_file_path(dir, false, key, cf);
+                let clone_path = get_cf_clone_file_path(dir, false, key, cf);
+                prepare_sst_for_ingestion(&cf_path, &clone_path)?;
+
+                let db = options.db.as_ref();
+                let mut options = IngestExternalFileOptions::new();
+                options.move_files(true);
+                let handle = box_try!(get_cf_handle(db, cf));
+                let _timer = INGEST_SST_DURATION_SECONDS.start_coarse_timer();
+                db.ingest_external_file_cf(handle, &options, &[clone_path.to_str().unwrap()])?;
+            }
+        }
```

Comment:
Should set `self.applied` true here.

Reply:
`applied` is used for gc snapshots. It's should be set whatever it's applied successfully or not.